### PR TITLE
[CONSVC-2065] Set minimum coverage for combined unit and integration tests

### PR DESCRIPTION
**Pull request by Trinaa**
# Description
- Add a minimum combined unit and integration test coverage check for 95%
    - In the case of failure, Circle CI reports as follows:
    image
    image
- Output test results to file format compatible with CircleCI in order to take advantage of test insights
    - ref: [Collecting test data](https://circleci.com/docs/collect-test-data/)
- Add `test-results` directory to `.gitignore`
- Update documentation
# Issue(s)
#87 | [CONSVC-2065](https://mozilla-hub.atlassian.net/browse/CONSVC-2065)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 source = ["src"]
 branch = true
 relative_files = true
+data_file = "workspace/test-results/.coverage"
```
Comment (hackebrot) [inline pyproject.toml:5]: What does this do?

File: .gitignore [hunk 2]
```diff
@@ -38,6 +38,7 @@ pip-log.txt
 pip-delete-this-directory.txt
 
 # Unit test / coverage reports
+test-results/
```
Comment (hackebrot) [inline .gitignore:41]: Running `make test` generated this new directory for me locally:

```text
workspace
└── test-results
    ├── integration_results.xml
    └── unit_results.xml
```

Does it make sense to ignore `workspace/` and its subdirectories?

File: .circleci/config.yml [hunk 3]
```diff
@@ -82,17 +94,45 @@ jobs:
...
     steps:
       - checkout
+      - run:
+          name: Create Workspace
+          command: mkdir -p workspace
       - run:
           name: Unit tests
           command: make unit-tests
+      - store_test_results:
+          path: workspace/test-results
```
Comment (hackebrot) [inline .circleci/config.yml:104]: * in `Create Workspace` above we hardcode the path to the workspace: `workspace`
* the `Unit tests` step doesn't specify the test results dir: 🤔 
* the `store_test_results` step hardcodes the path to the test results dir: `workspace/test-results`

For this sequence of steps to work, `TEST_RESULT_DIR` in the `Makefile` **must** be set to `workspace/test-results`. Any changes to the default will cause the CI job to fail.

I recommend we hand over control over this path to the CircleCI config completely and set `TEST_RESULT_DIR` explicitly by setting an environment variable for the `Unit tests` and the `Integration tests` steps or for the entire job.

For reference: https://circleci.com/docs/env-vars/#environment-variable-usage-options

File: Makefile [hunk 4]
```diff
@@ -1,5 +1,7 @@
 APP_DIR := merino
 TEST_DIR := tests
+TEST_RESULT_DIR := "workspace/test-results"
```
Comment (hackebrot) [inline Makefile:3]:
```suggestion
TEST_RESULTS_DIR := "workspace/test-results"
```

**Review by hackebrot — CHANGES_REQUESTED** [review]
Hi @Trinaa! 👋🏻 

This is great! I tested it out locally and it worked! 🚀 

I've added a few inline comments for your consideration.

File: pyproject.toml (hunk 1 above)
Comment (Trinaa) [reply pyproject.toml:5]: In the _coverage_ module, the `data_file` option specifies the coverage file name and output location. By default, it's value is `.coverage`. The _pytest_ and _coverage_ modules will reference the pyproject.toml configurations by default.
Source: https://coverage.readthedocs.io/en/6.5.0/cmd.html#combining-data-files-coverage-combine

File: .gitignore (hunk 2 above)
Comment (Trinaa) [reply .gitignore:41]: Change applied: [825f6a6](https://github.com/mozilla-services/merino-py/pull/118/commits/825f6a603c093d8f8b12a6dd3c3ffdf7807f2cee)

File: Makefile (hunk 4 above)
Comment (Trinaa) [reply Makefile]: Change applied: [c3ab97b](https://github.com/mozilla-services/merino-py/pull/118/commits/c3ab97b79d49efba9311cfe7db23da7e9d9f2216)

File: .gitignore (hunk 2 above)
Comment (ncloudioj) [inline .gitignore:41]: This can be dropped as the root dir `workspace` is already ignored.

File: .dockerignore [hunk 5]
```diff
@@ -16,3 +16,5 @@ venv
 merino/configs/.secrets.*
 book/
 profile.json
+workspace/
+test-results/
```
Comment (ncloudioj) [inline .dockerignore:20]: This can be dropped as the root dir `workspace` is already ignored.

File: .circleci/config.yml [hunk 6]
```diff
@@ -80,19 +92,53 @@ jobs:
...
+      TEST_RESULTS_DIR: workspace/test-results
     steps:
       - checkout
+      - run:
+          name: Create Workspace
+          command: mkdir -p workspace
       - run:
           name: Unit tests
-          command: make unit-tests
+          command: make unit-tests TEST_RESULTS_DIR=$TEST_RESULTS_DIR
```
Comment (ncloudioj) [inline .circleci/config.yml:104]:
```suggestion
          command: make unit-tests
          environment:
            TEST_RESULTS_DIR: $TEST_RESULTS_DIR
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -80,19 +92,53 @@ jobs:
...
+      TEST_RESULTS_DIR: workspace/test-results
     steps:
       - checkout
+      - run:
+          name: Create Workspace
+          command: mkdir -p workspace
       - run:
           name: Integration tests
-          command: make integration-tests
+          command: make integration-tests TEST_RESULTS_DIR=$TEST_RESULTS_DIR
```
Comment (ncloudioj) [inline .circleci/config.yml:123]:
```suggestion
          command: make integration-tests
          environment:
            TEST_RESULTS_DIR: $TEST_RESULTS_DIR
```

**Review by ncloudioj — APPROVED** [review]
Looks great, r+ wc.

File: .dockerignore (hunk 5 above)
Comment (Trinaa) [reply .dockerignore]: Change applied: [a23c9aa](https://github.com/mozilla-services/merino-py/pull/118/commits/a23c9aa39851599af311e9e8583eb1e97292e95c)

File: .gitignore (hunk 2 above)
Comment (Trinaa) [reply .gitignore]: Change applied: [a23c9aa](https://github.com/mozilla-services/merino-py/pull/118/commits/a23c9aa39851599af311e9e8583eb1e97292e95c)

File: .circleci/config.yml (hunk 6 above)
Comment (Trinaa) [reply .circleci/config.yml:104]: In an experiment, I wasn't able to make this change work. I think the environment variable isn't considered by the Makefile or perhaps the Makefile definition is overriding. As discussed offline, I'll leave the command as is.

File: .circleci/config.yml (hunk 7 above)
Comment (Trinaa) [reply .circleci/config.yml:123]: See above [response](https://github.com/mozilla-services/merino-py/pull/118#discussion_r1029823112)

File: .circleci/config.yml (hunk 3 above)
Comment (Trinaa) [reply .circleci/config.yml]: Please confirm if I applied the change as expected.
Change applied: [47d0549](https://github.com/mozilla-services/merino-py/pull/118/commits/47d05495508861d5d92657e587b448282e44352c)

File: .circleci/config.yml (hunk 6 above)
Comment (hackebrot) [reply .circleci/config.yml:104]: We probably need to use a Conditional Variable Assignment in `Makefile` in this case:
https://www.gnu.org/software/make/manual/make.html#Conditional-Assignment

```bash
TEST_RESULTS_DIR ?= "workspace/test-results"
```

Comment (Trinaa) [reply .circleci/config.yml]: Change applied: [33b89b4](https://github.com/mozilla-services/merino-py/pull/118/commits/33b89b408eb093fb76cf580cdb61537d4354d0d2)

Note. I changed the 'store_test_results' path back to a literal. Not sure if it's possible to pass an environment variable.

**Review by hackebrot — APPROVED** [review]
LGTM! 🤖 📈 🚀

**Pull request merged**